src/Tools/dotnet-trace/CommandLine/Commands/MonitorCommand.cs

```diff
@@ -0,0 +1,302 @@
+                return ErrorCodes.ArgumentError;
+            }
+
+            // Build DiagnosticsClient
+            DiagnosticsClient diagnosticsClient;
+            if (ProcessLauncher.Launcher.HasChildProc)
+            {
+                try
+                {
+                    diagnosticsClient = ReversedDiagnosticsClientBuilder.Build(ProcessLauncher.Launcher, "dotnet-trace", 10);
+                }
+                catch (TimeoutException)
+                {
+                    Console.Error.WriteLine("Unable to start tracing session - the target app failed to connect to the diagnostics transport. This may happen if the target application is running .NET Core 3.1 or older versions. Attaching at startup is only available from .NET 5.0 or later.");
+                    if (!ProcessLauncher.Launcher.ChildProc.HasExited)
+                    {
+                        ProcessLauncher.Launcher.ChildProc.Kill();
+                    }
+                    return ErrorCodes.SessionCreationError;
+                }
+            }
+            else
+            {
+                diagnosticsClient = new DiagnosticsClient(processId);
+            }
+
+            EventPipeSession session = diagnosticsClient.StartEventPipeSession(providerCollection);
+            if (ProcessLauncher.Launcher.HasChildProc)
+            {
+                diagnosticsClient.ResumeRuntime();
+            }
+
+            ManualResetEvent shouldExit = new ManualResetEvent(false);
+            ct.Register(() => shouldExit.Set());
+
+            Task parserTask = Task.Run(() =>
+            {
+                try
+                {
+                    EventPipeEventSource source = new EventPipeEventSource(session.EventStream);
+
+                    foreach (IDiagnosticProfileHandler handler in GetAllHandlers(providerCollection))
+                    {
+                        handler.AddHandler(source);
+                    }
+                    source.Process();
+                }
+                catch (Exception ex)
+                {
+                    Debug.WriteLine(ex.ToString());
+                }
+                finally
+                {
+                    shouldExit.Set();
+                }
+            });
+
+            while(!shouldExit.WaitOne(250))
+            {
+                while (true)
+                {
+                    if (shouldExit.WaitOne(250))
+                    {
+                        StopSession(session);
+                        return 0;
+                    }
+                    if (Console.KeyAvailable)
+                    {
+                        break;
+                    }
+                }
+                ConsoleKey cmd = Console.ReadKey(true).Key;
+                if (cmd == ConsoleKey.Q)
+                {
+                    StopSession(session);
+                    break;
+                }
+            }
+
```

Comment:
Should we `await parserTask` here to make sure it's completed before exit?

Reply:
By the time we get here, we either excited out of parserTask (since shouldExit is also set by the parserTask when it exits), or the user stopped the session. I think it'd be redundant to put in the await here?